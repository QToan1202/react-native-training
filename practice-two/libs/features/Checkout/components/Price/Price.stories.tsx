import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Price } from '../index'

export default {
  title: 'components/Price',
  component: Price,
} as ComponentMeta<typeof Price>

const Template: ComponentStory<typeof Price> = (args) => <Price {...args} />

export const Default = Template.bind({})
Default.args = {
  data: [
    { id: 1, label: 'price (1 item)', value: '$25' },
    { id: 2, label: 'delivery fee', value: 'info' },
  ],
  total: 25,
}
