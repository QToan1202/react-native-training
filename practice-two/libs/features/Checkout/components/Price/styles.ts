import { StyleSheet } from 'react-native'

import { COLORS, FONT_FAMILY } from '@practice-two/shared/constants'

const styles = StyleSheet.create({
  text: {
    fontFamily: FONT_FAMILY.MONTSERRAT[500],
    lineHeight: 20,
    color: COLORS.BLACK,
    textTransform: 'capitalize',
  },
  title: {
    paddingVertical: 14,
    fontFamily: FONT_FAMILY.MONTSERRAT[600],
    fontSize: 18,
  },
})

export default styles
