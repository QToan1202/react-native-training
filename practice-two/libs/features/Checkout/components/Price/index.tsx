import { memo, useMemo } from 'react'
import isEqual from 'react-fast-compare'
import { Dimensions } from 'react-native'
import { Separator, XStack, YStack, YStackProps } from 'tamagui'

import Heading from '@components/Heading'
import Paragraph from '@components/Paragraph'
import { ICart } from '@types'

import styles from './styles'

export type PriceProps = {
  data: ICart[]
  deliveryFee: number
  header?: string
  footer?: string
} & YStackProps

const Price = ({
  data,
  deliveryFee,
  header = 'price details',
  footer = 'total amount',
  ...rest
}: PriceProps) => {
  const calculatePrice = useMemo(
    () =>
      data.reduce((previousValue: number, { price, discountPrice, quantity }: ICart) => {
        if (discountPrice !== 0) return previousValue + discountPrice * quantity

        return previousValue + price * quantity
      }, 0),
    [data]
  )

  return (
    <YStack
      paddingHorizontal="$space.3"
      paddingBottom="$space.5"
      backgroundColor="$color.white"
      {...rest}
    >
      <Heading
        fontWeight="$3"
        fontSize="$3"
        color="$color.black"
        textTransform="capitalize"
        content={header}
      />
      <XStack paddingTop="$space.3" alignItems="center" justifyContent="space-between">
        <YStack flex={1} space={10}>
          <XStack justifyContent="space-between">
            <Paragraph
              style={styles.text}
              content={`Price (${data.length} ${data.length > 1 ? 'items' : 'item'})`}
            />
            <Paragraph style={styles.text} content={`$${calculatePrice}`} />
          </XStack>
          <XStack justifyContent="space-between">
            <Paragraph style={styles.text} content="Delivery fee" />
            <Paragraph style={styles.text} content={`$${deliveryFee}`} />
          </XStack>
        </YStack>
      </XStack>
      <Separator
        alignSelf="stretch"
        marginTop="$space.5"
        marginLeft="$space.-3.5"
        borderBottomColor="$color.divider"
        width={Dimensions.get('window').width}
      />
      <XStack alignItems="center" justifyContent="space-between">
        <Heading style={[styles.text, styles.title]} content={footer} />
        <Heading
          style={[styles.text, styles.title]}
          fontWeight="$4"
          content={`$${calculatePrice + deliveryFee}`}
        />
      </XStack>
    </YStack>
  )
}

export default memo(Price, isEqual)
