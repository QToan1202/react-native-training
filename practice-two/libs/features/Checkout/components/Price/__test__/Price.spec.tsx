import { render, screen } from '@testing-library/react-native'

import { CHECKOUT } from '@constants'

import Price from '../index'

describe('Testing Price component', () => {
  beforeEach(() => {
    render(
      <Price
        data={CHECKOUT.PRICE_DETAILS}
        header="price details"
        footer="totals"
        total={25}
        testID="price-details"
      />
    )
  })

  it('Render component successfully', () => {
    const component = render(
      <Price data={CHECKOUT.PRICE_DETAILS} header="price details" footer="totals" total={25} />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Component must have right header and footer title', () => {
    const header = screen.getByText(/price details/)
    const footer = screen.getByText(/totals/)

    expect(header).toBeOnTheScreen()
    expect(footer).toBeOnTheScreen()
  })
})
