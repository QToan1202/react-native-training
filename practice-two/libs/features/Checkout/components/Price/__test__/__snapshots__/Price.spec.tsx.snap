// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Price component Render component successfully 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
        "paddingBottom": 15,
        "paddingHorizontal": 15,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#fff",
          "fontFamily": "Montserrat_700Bold",
          "fontSize": 24,
        },
        false,
        [
          {
            "color": "#000",
            "fontFamily": "Montserrat_500Medium",
            "lineHeight": 20,
            "textTransform": "capitalize",
          },
          {
            "fontFamily": "Montserrat_600SemiBold",
            "fontSize": 18,
            "paddingVertical": 14,
          },
        ],
      ]
    }
  >
    price details
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        {
          "rowGap": 10,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
          ]
        }
      >
        price (1 item)
      </Text>
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
          ]
        }
      >
        delivery fee
      </Text>
    </View>
    <View
      style={
        [
          {
            "rowGap": 10,
          },
          {
            "alignItems": "flex-end",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
          ]
        }
      >
        $25
      </Text>
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
          ]
        }
      >
        info
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "borderBottomColor": "#0000001a",
        "borderBottomWidth": 0.5,
        "marginLeft": -14,
        "marginTop": 24,
        "width": 750,
      }
    }
  />
  <View
    style={
      [
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          false,
          [
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
            {
              "fontFamily": "Montserrat_600SemiBold",
              "fontSize": 18,
              "paddingVertical": 14,
            },
          ],
        ]
      }
    >
      totals
    </Text>
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          false,
          [
            {
              "color": "#000",
              "fontFamily": "Montserrat_500Medium",
              "lineHeight": 20,
              "textTransform": "capitalize",
            },
            {
              "fontFamily": "Montserrat_600SemiBold",
              "fontSize": 18,
              "paddingVertical": 14,
            },
            {
              "fontFamily": "Montserrat_700Bold",
            },
          ],
        ]
      }
    >
      $25
    </Text>
  </View>
</View>
`;
